### ISSUE TITLE
POIs from POI format v3 files not shown

### ISSUE THREAD

**Issue by JFritzle** [Maproom/qmapshack#659]
### Describe the bug
While POI old format files are still working, POIs from POI v3 format files found at locations mentioned by QMS Wiki page [DocGisItemsPOI](https://github.com/Maproom/qmapshack/wiki/DocGisItemsPOI)
o [Mapsforge](http://download.mapsforge.org/pois/) and
o [GWDG](https://ftp.gwdg.de/pub/misc/openstreetmap/openandromaps/pois/ORUX_V3/) 
are no longer shown.

### What have you done to circle down the problem?
[comment]: # (Did you compare to previous versions? What versions? Did you check if others have the same problem?)
Currently, POI old format files are still available at [GWDG](https://ftp.gwdg.de/pub/misc/openstreetmap/openandromaps/pois/ORUX_old/), which are processed correctly as well as in recent QMS version 1.17.1 as in previous QMS versions. POIs from POI v3 format files are not shown even in recent and previous QMS versions.

### To Reproduce
[comment]: # (List the steps to reproduce the behavior)

1. Download e.g. Mapsforge POI file [berlin.poi](http://download.mapsforge.org/pois/europe/germany/berlin.poi) or POI file [Berlin_oam.osm.poi](https://ftp.gwdg.de/pub/misc/openstreetmap/openandromaps/pois/ORUX_V3/germany/Berlin.PoiV3.zip).
2. Activate POI in QMS
3. Select one or more POI categories
4. Notice that no POI is shown

### Expected behavior
[comment]: # (A clear and concise description of what you expected to happen)
As POI old format files are no longer available at [Mapsforge](http://download.mapsforge.org/pois/) and may become unavailable at [GWDG](https://ftp.gwdg.de/pub/misc/openstreetmap/openandromaps/pois/) in future, QMS should be able to process recent POI v3 format files too.

### Screenshots
[comment]: # (If applicable, add screenshots to help explain your problem)
POI old format file [Berlin](https://ftp.gwdg.de/pub/misc/openstreetmap/openandromaps/pois/ORUX_old/germany/Berlin.PoiOld.zip), showing railway stations:
![Berlin Railway Stations (old version POI file)](https://github.com/Maproom/qmapshack/assets/62614244/9cb731c0-b7c0-45a4-80de-f086bce2051d)

POI format v3 file [Berlin](https://ftp.gwdg.de/pub/misc/openstreetmap/openandromaps/pois/ORUX_V3/germany/Berlin.PoiV3.zip), **not** showing railway stations:
![Berlin Railway Stations (new version POI file)](https://github.com/Maproom/qmapshack/assets/62614244/c534cb75-fe11-4d6d-a86f-6e4f33ceb5a9)

### Attachments
[comment]: # (Can you provide an example file to trigger the problem?)
-

### Tracebacks
[comment]: # (Add your backtrace below if you have one. If QMapshack crashes the fastest way to get help is a backtrace.
For Linux see: https://github.com/Maproom/qmapshack/wiki/TroubleShooting#create-a-backtrace-of-a-crash-on-linux.
For Windows it would need a debug build and running QMapshack in Visual Studio.)
-

### Desktop
[comment]: # (please complete the following information)
 - OS: Windows
 - QMapShack Version: 1.17.1 and 1.16.1

### Additional context
[comment]: # (Add any other context about the problem here)
Internal data structure of recent Mapsforge/OpenAndroMaps POI files has been changed to format v3 as described [here](https://www.openandromaps.org/en).

**Comment by mitxel-m** [comment]
I have also been looking at this issue in the last few days.I bring here what I have been checking and hope it helps as a starting point.

The internal structure of the `.poi` file has changed, it is no longer using an r-tree, and the structure of the poi_index table is different. 

##### In V2
![POI_V2_scheme](https://github.com/Maproom/qmapshack/assets/39693664/4c2e00a1-2b18-4b12-ac1b-9c56d359ba86)

where `poi_index_node`, `poi_index_parent`,`poi_index_rowid`, are internales tables used by r-tree

##### In V3
![POI_v3_scheme](https://github.com/Maproom/qmapshack/assets/39693664/ebca970b-efe4-425a-8585-94ab2d67d6df)

In `poi_index`table columns `lat`and `lon`are indexed, but there is no r-tree
structure.

Since in QMS a query is made based on the columns `minLat`, `maxLat`, `minLon`, `maxLon` from V2, and they no longer exist in V3, then the query does not return any results.


In principle the solution is to modify the query according to the new `lat` and `lon` column names. I have already tried this and it works, but with two factors that we should study better:


1. By doing this, the poi V2 files cannot be read by QMS anymore. So if we want compatibility with V2 and V3 we would first have to check which is the version of the POI file, (in the metadata table), and consequently launch one query or another. ... Or else decide to jump to V3 and drop V2. 


2. It seems to me that after this change QMS takes much longer to load the  V3 POI than the V2, possibly because it is not taking advantage of the index of the poi_index table. For now I don`t know if this has a solution or how to improve it. (needs further study)

I've opened a branch here ( https://github.com/mitxel-m/qmapshack/tree/QMS-659)  that already contains the changes to the query so you can try it out. It`s a draft and only a starting point.


### Alternatively 

There is another way, which is possible but maybe not very elegant.
Using a sql script we can convert a v3 file to v2 very quickly.  I have already made and tested a script and it works for me:

````
BEGIN TRANSACTION;
ALTER TABLE poi_index RENAME TO poi_index_ori;
CREATE VIRTUAL TABLE poi_index USING rtree(id, minLat, maxLat, minLon, maxLon);
INSERT INTO poi_index select id, lat as minLat, lat as maxLat, lon as minLon, lon as maxLon from poi_index_ori where poi_index_ori.id is not null;
DROP TABLE poi_index_ori;
UPDATE metadata SET value = 2 where name = 'version';
COMMIT;
VACUUM;
````

So you can put this in a ' POI_V3toV2.sql' text file ([attached here](https://github.com/Maproom/qmapshack/files/13786436/POI_V3toV2.sql.zip)) and then in a terminal do:
````
sqlite3 Spain-Portugal.poi < POI_V3toV2.sql

````

For reference It takes a few seconds to convert Canary_islands, and about 90 seconds to convert Spain-Portugal (over 1.777.000 items). With this we would keep the current speed loading pois and no change is necessary in the code, but it requires the user to convert v3 files in the future, so it would be a last chance remedy, or in case we see that with V3 QMS slows down a lot.

**Comment by mitxel-m** [comment]
I have added a new commit in my [branch QMS-659]( https://github.com/mitxel-m/qmapshack/tree/QMS-659). With these latest changes it is possible to load POI files in both v2 and v3 versions. QMS first checks the file version by looking at the metadata table and then prepares the query according to that version.

So far everything is fine, but for now QMS works much faster using POI v2 files than using v3 files. Using a file for a small region (eg: Canary islands 59.000 items) the differences in load time are very small from v2 to v3, however if the POI file is large (eg: the whole Spain and Portugal 1.700.000 items) differences are notorius especially when:

- more than one category is checked at the same time
- the zoom is more distant, even if that area does not include new items.

As a reference, with a zoom adjusted to the island of Mallorca, and marking the whole tree of the `Accomadation` category V2 takes 4 seconds to load, and v3 takes 30 seconds. If I zoom out a bit more the difference increases (5 seconds vs 55 seconds).

This leads me to think that the current code works very well with the v2 files that have an r-tree, but I don't know if we will be able to get the same efficiency with the v3 files without making deep changes in the code, and in that aspect I'm a bit lost. @pingurus did the development for using POIs in QMS and his help would be very much appreciated at this point.

I have also tried to modify the query for the v3 files trying to take advantage of the built-in index, but without success.

It's also worth remembering that converting a v3 to v2 is relatively easy and fast.

To summarize we are at this point:
- with the last changes on [branch QMS-659]( https://github.com/mitxel-m/qmapshack/tree/QMS-659)  you can load v2 and v3.
- QMS works slower with v3 and to improve it would require more changes in the code.
- alternatively it is possible to convert v3 files to v2 with a script.

**Comment by d029940** [comment]
My experiences are:

[https://ftp.gwdg.de/pub/misc/openstreetmap/openandromaps/mapsV4/](https://ftp.gwdg.de/pub/misc/openstreetmap/openandromaps/mapsV4/) are working.
[https://ftp.gwdg.de/pub/misc/openstreetmap/openandromaps/mapsV5/](https://ftp.gwdg.de/pub/misc/openstreetmap/openandromaps/mapsV5/) are not working.

I wonder, which is the preferred version.

Tested on Windows 11 and MacOS 14 (Sonoma). On both OS with QMS V 1.17.1

**Comment by JFritzle** [comment]
The **MAP format** of the current OpenAndroMaps has been V5 for some time. Maps at https://ftp.gwdg.de/pub/misc/openstreetmap/openandromaps/mapsV5/ are bundled with POI files. As of December 2023, the **POI format** of the bundled POI files has changed from V2 to V3 format. See [announcement](https://www.openandromaps.org/en). 

For QMS version 1.17.1 and before, these choices are currently available:
- download outdated V4 format MAP files bundled with outdated V2 format POI files
- download recent V5 format MAP files bundled with recent V3 format POI files and convert V3 format POI files to V2 format POI files using SQL script mentioned above.
- as long as V2 format POI files are still supplied, seperately download recent V2 format POI files from https://ftp.gwdg.de/pub/misc/openstreetmap/openandromaps/pois/ORUX_old/

**Comment by kiozen** [comment]
I would vote for converting the new format into the old one unless someone volunteers to analyze the new one to find a more optimized way to read it. And is doing the implementation.

Users are greedy collectors. They hoard every bit of data they can get. And they expect that software can cope with that senseless amount of data. Therefore a performance hit is not acceptable. The POI stuff is already slow with the old format.

**Comment by JFritzle** [comment]
Installing a new POI file is not an everyday task. Therefore, the one-time conversion from the V3 format to the V2 format is a reasonable effort.

In my opinion, it is acceptable to require POI files in V2 format as long as there is no implementation that speeds up reading POI files in V3 format. But when trying to activate a V3 POI file, a message should pop up telling that this format is not supported. Including a hint on how the V3 format can be converted easily to the V2 format using an sqlite3 script. 

It would of course be very nice if the user could decide within QMS to convert the V3 format file into the V2 format on the fly, assuming write permission.

**Comment by d029940** [comment]
As for conversion:
- On Linux and MacOS I found it an easy process.
- On Windows, you need to install sqlite first.
- I think Linux users are more tied to system administration than MacOS- / Windows users.

Since it is not too difficult, it is a matter of documentation. Especially that you can download up-to-date POIs in v2 format.
I didn't know, that this exists, and so will a lot of users.

Sure, ultimately, support of the V3 format out of the box, would be very nice.
But I fully understand the issue that someone needs to develop it. And we are all volunteers working in our free time.

**Comment by JFritzle** [comment]
On Windows, you will find a usable `sqlite3.exe` in QMS installation folder, at least with QMS v1.17.1. Therefore no installation is required. Using sqlite3 script `POI_V3toV2.sql` mentioned above, simple Windows command 
 `"C:\Program Files\QMapShack\sqlite3.exe" poi_v3_file.poi < POI_V3toV2.sql` 
will successfully convert POI format V3 file `poi_v3_file.poi` to POI format V2.

**Comment by d029940** [comment]
Even better.

**Comment by kkarsten62** [comment]
I just updated my POI file.

Thanks to @mitxel-m . The sql script `POI_V3toV2.sql` works out of the box on my Linux Fedora 41 box.

I have created a small Linux bash script to simplify to conversion process. Feel free to use it.
 
1 . Copy the following code to a file named `build_POI.sh` to your QMS POI folder
```
#!/bin/bash

# Related to QMS issue: https://github.com/Maproom/qmapshack/issues/659

url='https://ftp.gwdg.de/pub/misc/openstreetmap/openandromaps/pois/mapsforge/'

# Structure is:
# 'Subfolder to url,filename w/o extension'
datas=(
'europe/,Austria'
'europe/,Belgium'
'europe/,France-North'
'europe/,France-South'
'germany/,Germany-North'
'germany/,Germany-Mid'
'germany/,Germany-South'
'europe/,Italy-North'
'europe/,Italy-South'
'europe/,Luxembourg'
'europe/,Netherlands'
'europe/,Switzerland'
)

no="N"
read -p $'Delete existing POI files in this folder, y/N ? ' no
if [[ "$no" != "y" && "$no" != "Y" ]]; then
	exit
fi

for data in "${datas[@]}"
do
	poiFolder=$(echo "${data}" | awk -F',' '{ print $1 }')
	poiFile=$(echo "${data}" | awk -F',' '{ print $2 }')

	echo -e "\n\n --- Converting file: ${poiFile} ---"
	if [ -f ${poiFile}'_oam.osm.poi' ]; then
		rm ${poiFile}'_oam.osm.poi'
	fi

	wget ${url}${poiFolder}${poiFile}'.Poi.zip'

	unzip ${poiFile}'.Poi.zip'

	sqlite3 ${poiFile}'_oam.osm.poi' < POI_V3toV2.sql

	if [ -f ${poiFile}'.Poi.zip' ]; then
		rm ${poiFile}'.Poi.zip'
	fi
done
```
2. Make the sh script executable by `chmod 755 build_POI.sh`
3. Copy the sql script `POI_V3toV2.sql` to the same QMS POI folder
4. Adjust the `build_POI.sh' for your needed POI files, according the subfolder structure in

https://ftp.gwdg.de/pub/misc/openstreetmap/openandromaps/pois/mapsforge/

```
# Structure is:
# 'Subfolder to url,filename w/o extension'
datas=(
'europe/,Austria'
'europe/,Belgium'
'europe/,France-North'
'europe/,France-South'
'germany/,Germany-North'
...
)
```

5. Start the script by `./build_POI.sh`. Script will download the POI file and do the conversion from V3 to V2.
6. Result should be the related POI files `*_oam.osm.poi`.
7. In QMS reload your POI files.

Hope it helps.

**Comment by JFritzle** [comment]
I suggest, to replace
```
CPoiFilePOI::CPoiFilePOI(const QString& filename, CPoiDraw* parent)
    : IPoiFile(parent), filename(filename), loadTimer(new QTimer(this)) {
...
 // Open database here so it belongs to the right thread
  if (!QSqlDatabase::contains(filename + "_bbox")) {
    QSqlDatabase db = QSqlDatabase::addDatabase("QSQLITE", filename + "_bbox");
    db.setDatabaseName(filename);
    if (!db.open()) {
      qDebug() << "failed to open database" << db.lastError();
      isActivated = false;
      return;
    }
  }

  QSqlQuery query("SELECT value FROM main.metadata WHERE name='bounds'", QSqlDatabase::database(filename + "_bbox"));

  if (query.next()) {
    const QStringList& corners = query.value(0).toString().split(",");
    qreal top = corners[2].toDouble();
    qreal left = corners[1].toDouble();
    qreal width = corners[3].toDouble() - left;
    qreal height = corners[0].toDouble() - top;
    bbox = QRectF(left, top, width, height);
  } else {
    qDebug() << "failed to retrieve bounding box of " << filename;
    isActivated = false;
  }
...
}
```
by an additional check for POI version 2 and more meaningful output to qInfo channel instead of qDebug channel
```
CPoiFilePOI::CPoiFilePOI(const QString& filename, CPoiDraw* parent)
    : IPoiFile(parent), filename(filename), loadTimer(new QTimer(this)) {
...
 // Open database here so it belongs to the right thread
  if (!QSqlDatabase::contains(filename + "_bbox")) {
    QSqlDatabase db = QSqlDatabase::addDatabase("QSQLITE", filename + "_bbox");
    db.setDatabaseName(filename);
    if (!db.open()) {
      qInfo() << tr("File '%1' is no valid POI file,").arg(filename).toStdString().c_str();
      qInfo() << tr("failed to open database: ").toStdString().c_str() << db.lastError();
      isActivated = false;
      return;
    }
  }

  QSqlQuery query("SELECT value FROM main.metadata WHERE name='bounds'", QSqlDatabase::database(filename + "_bbox"));

  if (query.next()) {
    const QStringList& corners = query.value(0).toString().split(",");
    qreal top = corners[2].toDouble();
    qreal left = corners[1].toDouble();
    qreal width = corners[3].toDouble() - left;
    qreal height = corners[0].toDouble() - top;
    bbox = QRectF(left, top, width, height);
  } else {
    qInfo() << tr("File '%1' is no valid POI file,").arg(filename).toStdString().c_str();
    qInfo() << tr("failed to retrieve bounding box!").toStdString().c_str();
    isActivated = false;
  }

  if (isActivated) {

    QSqlQuery query("SELECT value FROM main.metadata WHERE name='version'", QSqlDatabase::database(filename + "_bbox"));

    QString version = "unknown";
    if (query.next()) {
      version = query.value(0).toString();
    }
    if (version != "2") {
      qInfo() << tr("Poi file '%1' is POI version %2.").arg(filename).arg(version).toStdString().c_str();
      qInfo() << tr("Only version 2 is supported! See https://github.com/Maproom/qmapshack/wiki/DocGisItemsPOI").toStdString().c_str();
      isActivated = false;
    }

  }
...
}
```

Benefit:

- Messages are shown even without running QMS with debug enabled.
- Messages tell reasons of POIs not being available for a given file.
- Messages give a hint what to do with POI version 3 files. This requires adding a paragraph to the wiki on how to backport a POI version 3 file to a POI version 2 file as described above.
- POI version 3 files are included in the list not containing any POI category and cannot be activated. An info message appears when attempting to activate. 

Messages output is looking like that:
```
2025-07-26 19:32:09.162 [info] Poi file '/mnt/hgfs/D/Landkarten/OpenAndroMaps/POI/Alps.poi' is POI version 3.
2025-07-26 19:32:09.162 [info] Only version 2 is supported! See https://github.com/Maproom/qmapshack/wiki/DocGisItemsPOI
```

If you agree, I will contribute this change.
